### PR TITLE
Preprocess image right after downloading and pycurl to download images

### PR DESCRIPTION
**Pull request by wanliAlex**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
feature

* **What is the current behavior?** (You can also link to an open issue here)
In `add_documents`, we download and cache the images until we send all the images to `vectorise`. 
This may cause memory issues as Marqo is caching too many images.

* **What is the new behavior (if this is a feature change)?**
We preprocess the images to tensors right after downloading. Marqo no long caches the full-sized images anymore.
This strategy significantly improves the memory efficiency of Marqo.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
no

* **Have unit tests been run against this PR?** (Has there also been any additional testing?)
no

* **Related Python client changes** (link commit/PR here)
no

* **Related documentation changes** (link commit/PR here)
no

* **Other information**:
no

* **Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: src/marqo/s2_inference/s2_inference.py [hunk 1]
```diff
@@ -160,6 +167,52 @@ def get_marqo_inference_cache() -> MarqoInferenceCache:
     return _marqo_inference_cache
 
 
+def is_preprocess_image_model(model_name: str, model_properties: dict = None) -> bool:
+    """Check if the model should be preloaded with an image preprocessor to preprocess image tensor_search module"""
+
+    validated_model_properties = validate_model_properties(model_name, model_properties)
```
Comment (farshidz) [inline src/marqo/s2_inference/s2_inference.py]: The Model class already performs this validation

File: src/marqo/s2_inference/s2_inference.py [hunk 2]
```diff
@@ -160,6 +167,52 @@ def get_marqo_inference_cache() -> MarqoInferenceCache:
...
+        InternalError: If the device is not set.
+        InternalError: If the model is not a model that requires preload image preprocessor.
+    """
+    if not device:
+        raise InternalError(message=f"vectorise (internal function) cannot be called without setting device!")
+
+    if not is_preprocess_image_model(model_name, model_properties):
+        raise InternalError(message=f"Model {model_name} is not a model that requires preload image preprocessor.")
+
+    validated_model_properties = validate_model_properties(model_name, model_properties)
```
Comment (farshidz) [inline src/marqo/s2_inference/s2_inference.py]: Model class validates

File: src/marqo/s2_inference/s2_inference.py (hunk 1 above)
Comment (wanliAlex) [reply src/marqo/s2_inference/s2_inference.py]: done

File: src/marqo/s2_inference/s2_inference.py (hunk 2 above)
Comment (wanliAlex) [reply src/marqo/s2_inference/s2_inference.py]: done

**Review by farshidz — APPROVED** [review]

**Pull request merged**